### PR TITLE
Issue #SH-401: added reissue certificate for in portal backend

### PR DESCRIPTION
**Pull request by swetha-aggidevara**

### CODE REVIEW

File: src/app/routes/certRegRoutes.js [hunk 1]
```diff
@@ -5,13 +5,49 @@ const certRegURL = envHelper.LEARNER_URL
 const reqDataLimitOfContentUpload = '50mb'
 const proxy = require('express-http-proxy')
 const logger = require('sb_logger_util_v2')
-
+_ = require('lodash')
```
Comment (sudip-mukherjee) [inline src/app/routes/certRegRoutes.js:8]: missing const before _

File: src/app/routes/certRegRoutes.js [hunk 2]
```diff
@@ -5,13 +5,49 @@ const certRegURL = envHelper.LEARNER_URL
...
+const bodyParser = require('body-parser');
+const { getUserCertificates } = require('./../helpers/certHelper');
 var certRegServiceApi = {
-  searchCertificate : 'certreg/v1/certs/search'
+  searchCertificate : 'certreg/v1/certs/search',
+  reIssueCert: '/certreg/v1/user/search',
```
Comment (sudip-mukherjee) [inline src/app/routes/certRegRoutes.js:13]: make it reIssueCertificate

File: src/app/helpers/certHelper.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+      request: {
+        _source: [],
+        query: {
+          bool: {
+            must: [
+              {
+                match_phrase: {
+                  'recipient.id': _.get(userData, 'identifier'),
+                },
+              },
```
Comment (sudip-mukherjee) [inline src/app/helpers/certHelper.js:25]: why do we need to add these many commas (,) ? we only have one match_phrase, right ?

File: src/app/routes/certRegRoutes.js [hunk 4]
```diff
@@ -32,4 +38,36 @@ module.exports = function (app) {
...
-};
+
+    let courseId;
+    app.post(certRegServiceApi.reIssueCertificate,
+      bodyParser.json({ limit: '10mb' }),
+      permissionsHelper.checkPermission(),
+      proxy(certRegURL, {
+        proxyReqOptDecorator: proxyUtils.decorateRequestHeaders(),
+        proxyReqPathResolver: function (req) {
+          logger.debug({msg: `${certRegServiceApi.reIssueCertificate} is called with request: ${JSON.stringify(_.get(req, 'body'))}`});
```
Comment (vinukumar-vs) [inline src/app/routes/certRegRoutes.js:49]: this should be logger.info method
"Certificate re-issue has called."

Dont add debug logs. In error log we should add request URL & payload as well.

File: src/app/routes/certRegRoutes.js [hunk 5]
```diff
@@ -32,4 +38,36 @@ module.exports = function (app) {
         }
       }
     }))
-};
+
+    let courseId;
+    app.post(certRegServiceApi.reIssueCertificate,
+      bodyParser.json({ limit: '10mb' }),
+      permissionsHelper.checkPermission(),
```
Comment (vinukumar-vs) [inline src/app/routes/certRegRoutes.js:45]: We should allow only admin can do these actions. Restrict for any other roles.

Comment (vinukumar-vs) [reply src/app/routes/certRegRoutes.js:45]: You should add this in routes into whitelist api's list with the role checks

File: src/app/routes/certRegRoutes.js [hunk 6]
```diff
@@ -5,9 +5,15 @@ const certRegURL = envHelper.LEARNER_URL
...
 const proxy = require('express-http-proxy')
 const logger = require('sb_logger_util_v2')
+const _ = require('lodash')
+const bodyParser = require('body-parser');
+const { getUserCertificates } = require('./../helpers/certHelper');
+
 
 var certRegServiceApi = {
-  searchCertificate : 'certreg/v1/certs/search'
+  searchCertificate : 'certreg/v1/certs/search',
```
Comment (vinukumar-vs) [inline src/app/routes/certRegRoutes.js:14]: Only admin role should be allowed to do these actions

File: src/app/routes/certRegRoutes.js [hunk 7]
```diff
@@ -32,4 +38,36 @@ module.exports = function (app) {
...
+            const data = JSON.parse(proxyResData.toString('utf8'));
+            logger.info({msg: `getUserCertificates() is calling from certRegRoutes `});
+            const certificates = await getUserCertificates(data, courseId);
+            if (data) {
+            data.result.response = certificates;
+            return data;
+            }
+            else return proxyUtils.handleSessionExpiry(proxyRes, proxyResData, req, res, data);
+          } catch (err) {
+            logger.error({msg: `Error occured while searching userData with: ${certRegServiceApi.searchUser}, Error: ${err}`});
```
Comment (vinukumar-vs) [inline src/app/routes/certRegRoutes.js:65]: log request payload as well here.

File: src/app/routes/certRegRoutes.js (hunk 6 above)
Comment (vinukumar-vs) [inline src/app/routes/certRegRoutes.js:14]: difference in 
searchCertificate & reIssueCertificate

File: src/app/routes/certRegRoutes.js (hunk 5 above)
Comment (swetha-aggidevara) [reply src/app/routes/certRegRoutes.js:44]: Currently Dashboard can be seen only by MENTORS (that to if he has creator batches for that course) . SO Admin means => ("ORG_ADMIN","SYSTEM_ADMINISTRATION") or (the creator of the batch  (MENTOR)). ??

File: src/app/client/src/app/modules/dashboard/components/re-issue-certificate/re-issue-certificate.component.ts [hunk 8]
```diff
@@ -1,20 +1,138 @@
...
   selector: 'app-re-issue-certificate',
   templateUrl: './re-issue-certificate.component.html',
   styleUrls: ['./re-issue-certificate.component.scss']
 })
 export class ReIssueCertificateComponent implements OnInit, OnDestroy {
+  courseId: string;
+  @Input() userName;
+  userData: IUserCertificate;
+  certList: Array<{}>;
+  disableBtn = false;
```
[the remark this answers is not in the document]
Comment (swetha-aggidevara) [reply src/app/client/src/app/modules/dashboard/components/re-issue-certificate/re-issue-certificate.component.ts:20]: If it is not disabled then user can click search btn many
 times

File: src/app/client/src/app/modules/dashboard/components/re-issue-certificate/re-issue-certificate.component.ts [hunk 9]
```diff
@@ -1,20 +1,138 @@
...
+  instance: string;
```
[the remark this answers is not in the document]
Comment (swetha-aggidevara) [reply src/app/client/src/app/modules/dashboard/components/re-issue-certificate/re-issue-certificate.component.ts:21]: in most of the places it is being used as instance

**Pull request merged**